README.md
# NAME

Cot::Plugin::Session - Cot framework Simple session plugin.

# SYNOPSIS

    use Cot;
    use Cot::Plugin qw/Session/;

    get '/sign/in' => sub {
        my $self = shift;
        my $sess = $self->session->load;
        my $id = $sess->get('id');
        $sess->set('id', 'yshibata') unless $id;
        $sess->bakecookie(path => '/',
                          expires => time + 24 * 60 * 60,);
        #code
    };

# DESCRIPTION

Cot::Plugin::Session is Cot framework Simple session plugin.
using $cot->__session__ namespace.
Session file stored to $ENV{TMPDIR} or /tmp as a YAML file.

# FUNCTIONS

## get

Retrieve value with key from session object.

    use Cot;
    use Cot::Plugin qw/Session/;

    get '/sign/in' => sub {
        my $self = shift;
        my $sess = $self->session->load;
        my $id = $sess->get('id');
        #code
    };

## set

Store value with key to session object.

    my $sess = $self->session->load;
    $sess->set('id', 'yshibata');

## delete

Delete session object.

    my $sess = $self->session->load;
    $sess->delete;

## bakecookie

Bake session id COOKIE to HTTP header.

    my $sess = $self->session->load;
    $sess->bakecookie(path => '/',
                      expires => time + 24 * 60 * 60,
                      domain => '.example.com',);

# LICENSE

Copyright (C) Yusuke Shibata

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

Yusuke Shibata <shibata@yusukeshibata.jp>


Changes
Revision history for Perl extension Cot-Plugin-Session

0.07 2013-07-21T10:14:21Z

   - change store dir behavior.

0.06 2013-05-20T01:56:11Z

   - load function updated.

0.05 2013-04-26T00:02:26Z

   - not die when delete failed.

0.04 2013-04-24T08:16:53Z

    - pod fixed.

0.03 2013-04-24T07:53:35Z

    - big bug fix.

0.02 2013-04-24T05:43:04Z

    - get/set bug fix.

0.01 2013-04-23T00:38:19Z

    - original version



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Cot-Plugin-Session',
    module_name     => 'Cot::Plugin::Session',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


lib/Cot/Plugin/Session.pm
package Cot::Plugin::Session;
use strict;
use warnings;
use 5.008005;
our $VERSION = "0.07";
use parent qw(Cot::Plugin);
use File::Spec;
use Digest::SHA1 ();
use YAML         ();
use Carp;
our $SESSIONID = 'sessid';
use constant STOREDIR => '_cotsession';

sub init {
    my ( $self, $cot ) = @_;
    $self->{_app} = $cot;
    $self->{_dir} = File::Spec->catdir( $ENV{TMPDIR} || '/tmp', STOREDIR );
    $cot->session($self);
}

sub load {
    my ( $self, $sessid ) = @_;
    $sessid ||= $self->{_app}->req->cookies->{$SESSIONID};
    $sessid = Digest::SHA1::sha1_hex( rand() . $$ . {} . time )
      unless ( $sessid && -f $self->_path($sessid) );
    my $path    = $self->_path($sessid);
    my $sessobj = {};
    $sessobj = YAML::LoadFile($path) if ( -f $path );
    new Cot::Plugin::Session::Object(
        {
            _app => $self->{_app},
            dir  => $self->_dir($sessid),
            path => $path,
            obj  => $sessobj,
            id   => $sessid,
        }
    );
}

sub _dir {
    my ( $self, $sessid ) = @_;
    File::Spec->catdir(
        $self->{_dir},
        substr( $sessid, 0, 2 ),
        substr( $sessid, 2, 2 )
    );
}

sub _path {
    my ( $self, $sessid ) = @_;
    File::Spec->catfile(
        $self->{_dir},
        substr( $sessid, 0, 2 ),
        substr( $sessid, 2, 2 ), $sessid
    );
}

package Cot::Plugin::Session::Object;
use File::Path qw/mkpath/;

sub new {
    my ( $class, $obj ) = @_;
    bless $obj, $class;
}

sub get {
    my ( $self, $k ) = @_;
    $self->{obj}->{$k};
}

sub set {
    my ( $self, $k, $v ) = @_;
    $self->{obj}->{$k} = $v;
    mkpath( $self->{dir} ) unless -e $self->{dir};
    YAML::DumpFile( $self->{path}, $self->{obj} );
}

sub delete {
    my $self = shift;
    unlink $self->{path} if ( -f $self->{path} );
}

sub bakecookie {
    my ( $self, %param ) = @_;
    $self->{_app}->res->cookies->{$Cot::Plugin::Session::SESSIONID} = {
        value => $self->{id},
        %param,
    };
}

1;
__END__

=encoding utf-8

=head1 NAME

Cot::Plugin::Session - Cot framework Simple session plugin.

=head1 SYNOPSIS

    use Cot;
    use Cot::Plugin qw/Session/;

    get '/sign/in' => sub {
        my $self = shift;
        my $sess = $self->session->load;
        my $id = $sess->get('id');
        $sess->set('id', 'yshibata') unless $id;
        $sess->bakecookie(path => '/',
                          expires => time + 24 * 60 * 60,);
        #code
    };

=head1 DESCRIPTION

Cot::Plugin::Session is Cot framework Simple session plugin.
using $cot->B<session> namespace.
Session file stored to $ENV{TMPDIR} or /tmp as a YAML file.

=head1 FUNCTIONS

=head2 get

Retrieve value with key from session object.

    use Cot;
    use Cot::Plugin qw/Session/;

    get '/sign/in' => sub {
        my $self = shift;
        my $sess = $self->session->load;
        my $id = $sess->get('id');
        #code
    };

=head2 set

Store value with key to session object.

    my $sess = $self->session->load;
    $sess->set('id', 'yshibata');

=head2 delete

Delete session object.

    my $sess = $self->session->load;
    $sess->delete;

=head2 bakecookie

Bake session id COOKIE to HTTP header.

    my $sess = $self->session->load;
    $sess->bakecookie(path => '/',
                      expires => time + 24 * 60 * 60,
                      domain => '.example.com',);

=head1 LICENSE

Copyright (C) Yusuke Shibata

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Yusuke Shibata E<lt>shibata@yusukeshibata.jpE<gt>

